src/generator/mod.rs

```diff
@@ -183,58 +189,28 @@ impl<'a, B: Backend<'a>, W: Write> Generator<'a, B, W> {
             .unwrap_or(&mut self.default_out)
     }
 
-    fn resolve(&mut self, url: &str) -> Result<Include> {
-        let context = self
-            .iter_stack()
-            .find_map(|se| match se {
-                StackElement::Context(context) => Some(context),
+    pub fn diagnostics(&mut self) -> &mut Diagnostics<'a> {
+        self.stack
+            .iter_mut()
+            .rev()
+            .filter_map(|state| match state {
+                StackElement::Context(_, diagnostics) => Some(diagnostics),
                 _ => None,
             })
-            .expect("no Context???");
-        self.resolver.resolve(context, url)
+            .next()
```

Comment:
Extract this chain into a `fn top_context(&mut self) -> &mut (_, _)` to
* make the invariant clear
* reuse it here and in `resolve`